File: SwiftUI/EdgeStudio/Components/ProfileViewer/ProfileOperatorCard.swift
import SwiftUI

/// One operator in the execution plan, rendered as a rounded card —
/// matched to the VS Code extension's profile page.
///
/// Layout:
///   - Header row: operator name (bold, monospaced) followed by the
///     solid `ProfileStatsBadges` chips.
///   - Optional attributes block: key/value rows for operator-specific
///     fields (`collection`, `alias`, `datasource`, `limit`, …). Keys are
///     dimmed, values bold. Values that are JSON documents (e.g.
///     `descriptor`) render as a syntax-highlighted code block.
///
/// Children are NOT drawn here — `ProfileCardListView` handles
/// recursion so this card stays a single self-contained unit.
struct ProfileOperatorCard: View {
    let node: QueryProfileOperator

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(alignment: .firstTextBaseline, spacing: 10) {
                Text(node.name)
                    .font(.body.weight(.bold).monospaced())
                ProfileStatsBadges(stats: node.stats)
                Spacer(minLength: 0)
            }

            if !node.attributes.isEmpty {
                attributesGrid
            }
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Color.secondary.opacity(0.05))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .strokeBorder(Color.secondary.opacity(0.25), lineWidth: 0.5)
        )
    }

    /// Two-column key/value list. Keys take a fixed leading column so
    /// values align across rows; values are bold and monospaced. A value
    /// that parses as a JSON object/array renders as a highlighted code
    /// block instead of a single-line blob.
    private var attributesGrid: some View {
        VStack(alignment: .leading, spacing: 4) {
            ForEach(node.attributes, id: \.key) { attr in
                HStack(alignment: .top, spacing: 12) {
                    Text(attr.key)
                        .font(.caption.monospaced())
                        .foregroundStyle(.secondary)
                        .frame(width: 110, alignment: .leading)
                    if let json = Self.prettyPrintedJSON(attr.value) {
                        Text(JSONSyntaxHighlighter.highlight(json))
                            .font(.caption.monospaced())
                            .textSelection(.enabled)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(8)
                            .background(
                                RoundedRectangle(cornerRadius: 6)
                                    .fill(Color.secondary.opacity(0.10))
                            )
                    } else {
                        Text(attr.value)
                            .font(.caption.weight(.semibold).monospaced())
                            .foregroundStyle(.primary)
                            .textSelection(.enabled)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                }
            }
        }
        .padding(.leading, 4)
    }

    /// Pretty-prints `value` when it is a JSON object or array; nil otherwise.
    static func prettyPrintedJSON(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("{") || trimmed.hasPrefix("["),
              let data = trimmed.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data),
              let pretty = try? JSONSerialization.data(
                  withJSONObject: object,
                  options: [.prettyPrinted, .sortedKeys]
              ),
              let string = String(bytes: pretty, encoding: .utf8) else { return nil }
        return string
    }
}

File: android/app/src/main/java/com/costoda/dittoedgestudio/ui/mainstudio/profile/ProfileOperatorCard.kt
package com.costoda.dittoedgestudio.ui.mainstudio.profile

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.costoda.dittoedgestudio.domain.model.QueryProfileOperator

/**
 * One operator in the execution plan — matched to the VS Code extension's profile
 * page: bold monospaced name followed by the solid stat chips, then dimmed-key /
 * bold-value attribute rows. Values that are JSON documents (e.g. `descriptor`)
 * render as a syntax-highlighted code block.
 *
 * Children are NOT drawn here — [ProfileCardListView] handles recursion.
 */
@Composable
fun ProfileOperatorCard(
    operator: QueryProfileOperator,
    modifier: Modifier = Modifier,
) {
    Surface(
        modifier = modifier.fillMaxWidth().testTag("ProfileOperatorCard_${operator.name}"),
        shape = RoundedCornerShape(8.dp),
        color = MaterialTheme.colorScheme.surfaceContainerLow,
        border = BorderStroke(0.5.dp, MaterialTheme.colorScheme.outlineVariant),
    ) {
        Column(modifier = Modifier.padding(12.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = operator.name,
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace,
                )
                if (operator.stats != null) {
                    Spacer(modifier = Modifier.width(10.dp))
                    ProfileStatsBadges(stats = operator.stats)
                }
            }

            if (operator.attributes.isNotEmpty()) {
                Spacer(modifier = Modifier.height(8.dp))
                Column(modifier = Modifier.fillMaxWidth().padding(start = 4.dp)) {
                    operator.attributes.forEach { (key, value) ->
                        AttributeRow(key = key, value = value)
                    }
                }
            }
        }
    }
}

@Composable
private fun AttributeRow(
    key: String,
    value: String,
    modifier: Modifier = Modifier,
) {
    Row(modifier = modifier.padding(vertical = 1.dp)) {
        Text(
            text = key,
            style = MaterialTheme.typography.labelSmall,
            fontFamily = FontFamily.Monospace,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.width(110.dp),
        )
        val json = prettyPrintedJson(value)
        if (json != null) {
            Surface(
                shape = RoundedCornerShape(6.dp),
                color = MaterialTheme.colorScheme.surfaceContainerHighest,
                modifier = Modifier.padding(top = 2.dp),
            ) {
                SelectionContainer {
                    Text(
                        text = JsonProfileHighlighter.highlight(
                            json,
                            keywordColor = ProfileSyntaxColors.keyword,
                            stringColor = ProfileSyntaxColors.string,
                        ),
                        style = MaterialTheme.typography.labelSmall,
                        fontFamily = FontFamily.Monospace,
                        modifier = Modifier.padding(8.dp),
                    )
                }
            }
        } else {
            SelectionContainer {
                Text(
                    text = value,
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace,
                )
            }
        }
    }
}
